src/dlls/mscoree/unixinterface.cpp

```diff
@@ -87,6 +87,80 @@ static LPCWSTR* StringArrayToUnicode(int argc, LPCSTR* argv)
     return argvW;
 }
 
+static void ExtractStartupFlagsAndConvertToUnicode(
+    const char** propertyKeys,
+    const char** propertyValues,
+    int* propertyCountRef,
+    STARTUP_FLAGS* startupFlagsRef,
+    LPCWSTR** propertyKeysWRef,
+    LPCWSTR** propertyValuesWRef)
+{
+    int propertyCount = *propertyCountRef;
+
+    LPCWSTR* propertyKeysW = new (nothrow) LPCWSTR[propertyCount];
+    ASSERTE_ALL_BUILDS(propertyKeysW != nullptr);
+
+    LPCWSTR* propertyValuesW = new (nothrow) LPCWSTR[propertyCount];
+    ASSERTE_ALL_BUILDS(propertyValuesW != nullptr);
+
+    // Extract the startup flags from the properties, and convert the remaining properties into unicode
+    STARTUP_FLAGS startupFlags =
+        static_cast<STARTUP_FLAGS>(
+            STARTUP_FLAGS::STARTUP_LOADER_OPTIMIZATION_SINGLE_DOMAIN |
+            STARTUP_FLAGS::STARTUP_SINGLE_APPDOMAIN |
+            STARTUP_FLAGS::STARTUP_CONCURRENT_GC);
+    int propertyCountW = 0;
+    for (int propertyIndex = 0; propertyIndex < propertyCount; ++propertyIndex)
+    {
+        auto SetFlagValue = [&](STARTUP_FLAGS flag)
+        {
+            if (strcmp(propertyValues[propertyIndex], "0") == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags & ~flag);
+            }
+            else if (strcmp(propertyValues[propertyIndex], "1") == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | flag);
+            }
+        };
+
+        if (strcmp(propertyKeys[propertyIndex], "CONCURRENT_GC") == 0)
```

Comment:
I am sorry I came late here, I have had vacation. It looks a bit confusing to me to use the lambda expression capturing the ambient locals here. It seems it would be easier to read if you created a function that had the signature like:
```C++
bool TryExtractFlagValue(
    const char* key, 
    const char* value, 
    const char* flagKeyName, 
    STARTUP_FLAGS flag,
    STARTUP_FLAGS* accumulatedFlags
    );
```
and used it like this:
``` C++
char* key = propertyKeys[propertyIndex];
char* value = propertyValues[propertyIndex];
if (!TryExtractFlagValue(key, value, "CONCURRENT_GC", STARTUP_CONCURRENT_GC, &startupFlags) &&
    !TryExtractFlagValue(key, value, "SERVER_GC", STARTUP_SERVER_GC, &startupFlags) &&
    !TryExtractFlagValue(key, value, "HOARD_GC_VM", STARTUP_HOARD_GC_VM, &startupFlags) &&
    !TryExtractFlagValue(key, value, "TRIM_GC_COMMIT", STARTUP_TRIM_GC_COMMIT, &startupFlags)
    )
{
    // This is not a CoreCLR startup flag, convert it to unicode and preserve it for returning
    propertyKeysW[propertyCountW] = StringToUnicode(key);
    propertyValuesW[propertyCountW] = StringToUnicode(value);
    ++propertyCountW;
}

...
...
...
```